lib/dependencies/DependencyHooks.tsx

```diff
@@ -0,0 +1,239 @@
+  T extends any[],
+  K extends { [I in keyof T]: DependencyKey<I> } = {
+    [I in keyof T]: DependencyKey<T[I]>
+  }
+>(
+    keys: [...K]
+  ) => {
+  const currentValues = useDependencyValuesContext()
+  return keys.map((key) => currentValues.get(key)) as T
+}
+
+export type SetDependencyValueProps<T> = {
+  forKey: DependencyKey<T>
+  value: T
+  children: ReactNode
+}
+
+/**
+ * Sets the dependency value for a particular key in a child context. This is most useful
+ * when you need to update the value just for a single dependency, if you need to update
+ * the values for multiple dependencies, use `UpdateDependencyValues` instead.
+ *
+ * Using this will override the default/pre-existing value of the specified key.
+ *
+ * ```tsx
+ * const key = createDependencyKey(1);
+ *
+ * const Component = () => {
+ *  // Child components will now get the value 2 when trying to access the value for the key.
+ *  return (
+ *    <SetDependencyValue forKey={key} value={2}>
+ *      {...}
+ *    </SetDependencyValue>
+ *  )
+ * }
+ * ```
+ */
+export const SetDependencyValue = <T, >({
+  forKey: key,
+  value,
+  children
+}: SetDependencyValueProps<T>) => {
+  const newValues = useUpdateCopiedDependencyValues((values) => {
+    values.set(key, value)
+  })
+  return (
+    <DependencyValuesContext.Provider value={newValues}>
+      {children}
+    </DependencyValuesContext.Provider>
+  )
+}
+
+export type UpdateDependencyValuesProps = {
+  update: (values: DependencyValues) => void
+  children: ReactNode
+}
+
+/**
+ * Updates an instance of `DependencyValues` specifically for the child context to use.
+ *
+ * Use this component when you need to set the values of 2 or more dependencies. If you
+ * only need to set the value of 1 dependency, use `SetDependencyValue`.
+ *
+ * Using this will override any of the values for the given dependency keys.
+ *
+ * ```tsx
+ * const key1 = createDependencyKey(1);
+ * const key2 = createDependencyKey("hello");
+ *
+ * const Component = () => {
+ *   // Child components will now receive 2 for key1 and "world" for key2
+ *   return (
+ *     <UpdateDependencyValues
+ *       update={(values) => {
+ *         values.set(key1, 2);
+ *         values.set(key2, "world")
+ *       }}
+ *     >
+ *       {...}
+ *     </UpdateDependencyValues>
+ *   )
+ * }
+ * ```
+ */
+export const UpdateDependencyValues = ({
+  update,
+  children
+}: UpdateDependencyValuesProps) => {
+  const newValues = useUpdateCopiedDependencyValues(update)
+  return (
+    <DependencyValuesContext.Provider value={newValues}>
+      {children}
+    </DependencyValuesContext.Provider>
+  )
+}
+
+const DependencyValuesContext = createContext(new DependencyValues())
+
+const useDependencyValuesContext = () => useContext(DependencyValuesContext)
+
+const useUpdateCopiedDependencyValues = (
+  update: (values: DependencyValues) => void
+) => {
+  const currentValues = useDependencyValuesContext()
+  return useMemo(() => {
```

Comment:
How does this useMemo work. The value would always be rerendered since the dependencies are objects

Reply:
This `useMemo` ensures that we only perform the transformations on the dependencies when we need to (ie. In most cases we only need 1 transform, so why transform on each render unless we need to perform a transform based on a component's state?). 

While yes, the underlying `DependencyValues` object is well... an object, it should still retain the same memory address throughout renders (unless a parent component changes it via another transform) which would ensure that `useMemo` doesn't continuously run.